### PR TITLE
Added a mode to print a package's Package.swift contents in TOML

### PR DESCRIPTION
**Pull request by czechboy0**
Making the first step in making the Package.swift usable by other tools by adding a way to print the its parsed contents in the TOML format. 

As there's no TOML parser outside of SwiftPM, to make this much more useful, we should do one of these two things
1. write a TOML <-> JSON converter, by using most of the TOML parsing logic in SwiftPM and releasing it as a separate project
2. add serialization into JSON to Package in addition to the existing TOML, making it easy to print it in JSON as well (a flag would be added)

I think both are viable options, 1. is better for the long-term, we wouldn't need to maintain two sets of similar serialization code in SwiftPM, but 2. is better to enable tool integration with SwiftPM quickly.

/cc @ddunbar @mxcl @aciidb0mb3r  

(I also added a help line for --show-dependencies, keeping it in this PR to avoid a guaranteed conflict)

### CODE REVIEW

File: Sources/swift-build/usage.swift [hunk 1]
```diff
@@ -26,6 +26,8 @@ func usage(_ print: (String) -> Void = { print($0) }) {
     print("  --fetch                        Fetch package dependencies")
     print("  --update                       Update package dependencies")
     print("  --generate-xcodeproj[=<path>]  Generates an Xcode project [-X]")
+    print("  --show-dependencies[=<mode>]   Print dependency graph (text|dot|json)")
```
Comment (aciidb0mb3r) [inline Sources/swift-build/usage.swift]: This should be in separate commit

**Comment by aciidb0mb3r** [comment]
I would prefer the output to be in JSON as it can be parsed by almost all clients. We should also add additional information to the output for eg result from transmute(), pkg-config flags etc. If Foundation get implemented soon then we can use the `NSJSONSerialization` to do this more easily provided it is working on Linux.

File: Sources/swift-build/usage.swift (hunk 1 above)
Comment (czechboy0) [inline Sources/swift-build/usage.swift]: Normally I would put it into a separate PR, but it's just one line and would conflict with the line I'm adding. So I just sneaked it in and mentioned it in the PR description.

**Comment by czechboy0** [comment]
I would also prefer for this to be JSON, and it's my plan to take it there, but I want to make these changed bit by bit, verifying the direction is good in the process. I wanted to merge this first, then work on the JSON solution (whatever it'll be from the two options I outlined above). I really like it was done this way for the showing dependencies feature, split into relatively small PRs instead of one mega PR that you have to fix conflicts on all the time 😊 

@aciidb0mb3r From those two solutions above, which one would you prefer?

**Comment by aciidb0mb3r** [comment]
Definitely second. I am thinking just have JSON for these kind of things which operates independent of the TOML convertor and let the TOML part stay only for internal use. Anyway lets see what others think.

File: Sources/swift-build/usage.swift (hunk 1 above)
Comment (bhargavg) [inline Sources/swift-build/usage.swift]: Oops missed it 😅

**Comment by czechboy0** [comment]
@aciidb0mb3r Okay, it does make sense if you consider all the other things we'll probably want to vend using JSON. I just wanted to avoid duplication of serialization, but if it's just this one case and opens up the door to vending more information easily, I'm definitely on board with that 😊

**Comment by ddunbar** [comment]
I haven't looked at the PR yet, but my hope here is that we can just move to serializing using JSON. I think having a command to dump the package as JSON makes sense.

Does anyone know if Foundation on Linux has a JSON parser yet?

File: Sources/swift-build/usage.swift [hunk 2]
```diff
@@ -26,6 +26,8 @@ func usage(_ print: (String) -> Void = { print($0) }) {
...
+    print("  --show-package                 Print the contents of Package.swift")
```
Comment (ddunbar) [inline Sources/swift-build/usage.swift]: I would be a little inclined to call this `--dump-package` to differentiate it from something a typical developer would use commonly.

I might also consider dropping it into a different section in the `--help`.

File: Sources/swift-build/main.swift [hunk 3]
```diff
@@ -99,6 +99,12 @@ do {
         let (rootPackage, externalPackages) = try fetch(opts.path.root)
         dumpDependenciesOf(rootPackage: rootPackage, mode: mode)
 
+    case .ShowPackage:
+        let root = opts.path.root
+        let manifest = try parseManifest(path: root, baseURL: root)
+        let out = manifest.package.toTOML()
+        print(out)
```
Comment (ddunbar) [inline Sources/swift-build/main.swift]: My vote is that we only ever define this feature in terms of dumping the Package as JSON. I think that is the only format we would want to support externally.

**Comment by ddunbar** [comment]
To give a little more context:

1. We used TOML originally because we had no parser for either format, and it was easier to write a TOML one.

2. We also at that time weren't sure that we would stick with the manifest being in Swift, and viewed TOML as a more readable syntax if we did have to fall back. I view the Swift-based manifest as very successful, at this point, and so this is now irrelevant.

3. I would rather use JSON as the internal transport because it is more standard and we could use off the shelf tools. The only hold up is having access to code that implements them. If Foundation on Linux has JSON deserialization, let's use that.

4. We may still want to keep the TOML parser around, as there is one other situation where I think TOML might be a better choice for authoring and that is for `.gitconfig` style files if we grow a config facility. Of course, we can also blow away the TOML parser and resurrect it if we ever decide we want it.

File: Sources/swift-build/usage.swift (hunk 2 above)
Comment (czechboy0) [inline Sources/swift-build/usage.swift]: Haha I originally called it `--dump-package`, but then changed it because I thought we were avoiding the word `dump`.

**Comment by czechboy0** [comment]
@ddunbar @aciidb0mb3r Thanks both, after reading your points I totally agree. I'll close this PR for now and come back with a proper JSON-based serialization.

@ddunbar To your question, yes Swift Foundation already has NSJSONSerialization implemented (as I found out today). I'll see if I can get things working this way.

**Comment by ddunbar** [comment]
Great! That will of course also require we have made more progress along the "integrating Foundation" front but that is moving along quickly.

**Comment by czechboy0** [comment]
Ah Foundation doesn't yet build? Is there a ticket I can watch?

**Comment by ddunbar** [comment]
I don't think we have a ticket per se, but see all the Foundation related PRs (including ones I merged this morning)

**Comment by ddunbar** [comment]
(cc @Bouke)

**Comment by Bouke** [comment]
Foundation will be available in SwiftPM after merging #321, which is currently being tested on the CI.
